src/main/java/seedu/duke/storage/FoodListDecoder.java

```diff
@@ -1,49 +1,233 @@
+     */
+    private static boolean hasValidFoodName(String detail) throws ReadStorageFileErrorException {
+        if (detail.equals("")) {
+            throw new ReadStorageFileErrorException("food name: " + detail);
+        }
+        return true;
+    }
+
+    /**
+     * This method checks whether the encoded <code>expiryDate</code> is valid.
+     *
+     * @param detail is the encoded expiryDate
+     * @return true if valid
+     * @throws ReadStorageFileErrorException when not valid
+     */
+    private static boolean hasValidExpiryDate(String detail) throws ReadStorageFileErrorException {
+        try {
+            LocalDate.parse(detail, formatter);
+        } catch (DateTimeParseException e) {
+            throw new ReadStorageFileErrorException("expiry date: " + detail);
+        }
+        if (isInvalidDate(detail)) {
+            throw new ReadStorageFileErrorException("expiry date: " + detail);
+        }
+        return true;
+    }
+
+    /**
+     * This method checks whether the encoded <code>quantity</code> is valid.
+     *
+     * @param detail is the encoded quantity
+     * @return true if valid
+     * @throws ReadStorageFileErrorException when not valid
+     * @throws NumberFormatException when detail cannot be interpreted as a number
+     */
+    private static boolean hasValidQuantity(String detail) throws ReadStorageFileErrorException {
+        try {
+            double value = Double.parseDouble(detail);
+            if (value < 0 || value > 9999) {
+                throw new ReadStorageFileErrorException("food quantity:" + detail);
+            }
+        } catch (NumberFormatException e) {
+            Ui.showError(INVALID_QUANTITY_ERROR);
+        }
+        return true;
+    }
+
+    /**
+     * This method checks whether the encoded <code>category</code> is valid.
+     *
+     * @param detail is the encoded category
+     * @return true if valid; false otherwise
+     */
+    private static boolean hasValidCategory(String detail) {
+        for (FoodCategory category : FoodCategory.values()) {
+            if (category.name().equals(detail)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * This method checks whether the encoded <code>unit</code> is valid.
+     *
+     * @param detail is the encoded unit
+     * @return true if valid; false otherwise
+     */
+    private static boolean hasValidUnit(String detail) {
+        for (Unit unit : Unit.values()) {
+            if (unit.name().equals(detail)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * This method checks whether date is valid, based on the concept of leap year.
+     *
+     * @param date is the date to be checked
+     * @return true if valid; false if otherwise
+     */
+    public static boolean isInvalidDate(String date) {
+        String[] splitString = date.split("/", 3);
+        String d = splitString[0];

```

Comment:
Although it is obviious that d = day, m = month, y = year, the variable name can be clearer too by using day, month and year.